File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    stack = []
    
    for num in number:
        # 스택의 마지막 숫자가 현재 숫자보다 작고 제거할 수 있는 k가 남아있으면 제거
        while stack and k > 0 and stack[-1] < num:
            stack.pop()
            k -= 1
        stack.append(num)
    
    # 아직 제거해야 할 k가 남아있다면 뒤에서부터 제거
    if k > 0:
        stack = stack[:-k]
    
    return ''.join(stack)

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        String answer = "";
        Stack<Character> stack = new Stack<>();
        int toRemove = k;
        
        for (int i = 0; i < number.length(); i++) {
            char digit = number.charAt(i);
            while(!stack.isEmpty() && stack.peek() < digit) {
                if (toRemove == 0) break;
                stack.pop();
                toRemove--;
            }
            stack.push(digit);
        }
        
        for (int i = 0; i < number.length() - k; i++) {
            answer += stack.get(i);
        }
        
        return answer;
    }
}